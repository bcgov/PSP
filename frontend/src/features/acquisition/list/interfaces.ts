export interface IAcquisitionFilter {
  acquisitionFileStatusTypeCode: string;
  acquisitionFileNameOrNumber: string;
  projectNameOrNumber: string;
}
