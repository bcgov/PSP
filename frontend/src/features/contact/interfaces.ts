export interface IContactFilter {
  summary: string;
  municipality: string;
  activeContactsOnly: boolean;
  searchBy: string;
}
