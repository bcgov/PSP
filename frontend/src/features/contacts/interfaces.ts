export interface IContactFilter {
  summary: string;
  municipality: string;
  activeContactsOnly: boolean;
  searchBy: string;
}

export enum ContactTypes {
  ORGANIZATION = 'O',
  INDIVIDUAL = 'P',
}
