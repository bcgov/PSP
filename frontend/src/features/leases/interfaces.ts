export interface ILeaseFilter {
  pinOrPid: string;
  lFileNo: string;
  searchBy: string;
  tenantName: string;
  programs: string[];
}
