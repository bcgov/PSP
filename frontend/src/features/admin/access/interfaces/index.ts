import { AccessRequestStatus } from 'constants/accessStatus';

export interface IAccessRequestModel {
  id: number;
  userId: number;
  businessIdentifierValue: string;
  firstName: string;
  surname: string;
  email: string;
  position: string;
  status: AccessRequestStatus;
  organization: string;
  role: string;
  note: string;
}
