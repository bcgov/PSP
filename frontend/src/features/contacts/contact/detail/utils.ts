import { IContactAddress } from 'interfaces/IContact';

import { AddressTypes } from './../../../../constants/addressTypes';

// the order of this array corresponds to the expected display order
const addressSortOrder = [AddressTypes.Mailing, AddressTypes.Residential, AddressTypes.Billing];

export const sortAddresses = (a1: IContactAddress, a2: IContactAddress) => {
  if (a2.addressType.id === a1.addressType.id) {
    return 0;
  }
  const a2Index = addressSortOrder.indexOf(a2.addressType.id as AddressTypes);
  if (a2Index !== -1 && a2Index < addressSortOrder.indexOf(a1.addressType.id as AddressTypes)) {
    return 1;
  }

  return -1;
};

export const fakeAddresses = [
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'BILLING',
      description: 'Billing address',
      isDisabled: false,
    },
    streetAddress1: 'Billing address',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6A 5G7',
  },
  {
    id: 1,
    rowVersion: 3,
    addressType: {
      id: 'MAILING',
      description: 'Mailing address',
      isDisabled: false,
    },
    streetAddress1: 'Mailing address',
    streetAddress2: 'Living in a van',
    streetAddress3: 'Down by the River',
    municipality: 'Hollywood North',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V6Z 5G7',
  },
  {
    id: 5,
    rowVersion: 1,
    addressType: {
      id: 'RESIDNT',
      description: 'Property address',
      isDisabled: false,
    },
    streetAddress1: 'Property address',
    streetAddress2: '',
    streetAddress3: '',
    municipality: 'vic',
    province: {
      provinceStateId: 1,
      provinceStateCode: 'BC',
      description: 'British Columbia',
    },
    country: {
      countryId: 1,
      countryCode: 'CA',
      description: 'Canada',
    },
    postal: 'V9A 0H6',
  },
];
