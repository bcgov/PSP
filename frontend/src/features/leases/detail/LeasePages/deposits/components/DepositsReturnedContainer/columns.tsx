import { Button } from 'components/common/form';
import { InlineFlexDiv } from 'components/common/styles';
import { ColumnWithProps, renderDate, renderMoney } from 'components/Table';
import Claims from 'constants/claims';
import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import {
  ILeaseSecurityDeposit,
  ILeaseSecurityDepositReturn,
  IOrganization,
  IPerson,
} from 'interfaces';
import { FaTrash } from 'react-icons/fa';
import { MdEdit } from 'react-icons/md';
import { CellProps } from 'react-table';
import styled from 'styled-components';

export class ReturnListEntry {
  public id: number;
  public depositTypeDescription: string;
  public terminationDate: string;
  public depositAmount: number;
  public claimsAgainst: number;
  public returnAmount: number;
  public returnDate: string;
  public personDepositHolder?: IPerson;
  public organizationDepositHolder?: IOrganization;

  public constructor(
    baseDeposit: ILeaseSecurityDepositReturn,
    parentDeposit: ILeaseSecurityDeposit,
  ) {
    this.id = baseDeposit.id || -1;
    if (baseDeposit.depositType.id === 'OTHER') {
      this.depositTypeDescription = (parentDeposit.otherTypeDescription || '') + ' (Other)';
    } else {
      this.depositTypeDescription = baseDeposit.depositType.description || '';
    }

    this.terminationDate = baseDeposit.terminationDate || '';
    this.depositAmount = parentDeposit.amountPaid;
    this.claimsAgainst = baseDeposit.claimsAgainst || 0;
    this.returnAmount = baseDeposit.returnAmount;
    this.returnDate = baseDeposit.returnDate || '';
    this.personDepositHolder = baseDeposit.personDepositReturnHolder;
    this.organizationDepositHolder = baseDeposit.organizationDepositReturnHolder;
  }
}

function renderPerson({ row: { original } }: CellProps<ReturnListEntry, string>) {
  return original.personDepositHolder?.fullName || '';
}

function depositActions(onEdit: (id: number) => void, onDelete: (id: number) => void) {
  return function({ row: { original, index } }: CellProps<ReturnListEntry, string>) {
    const { hasClaim } = useKeycloakWrapper();
    return (
      <StyledIcons>
        {hasClaim(Claims.LEASE_DELETE) && (
          <Button
            title="delete deposit return"
            icon={
              <FaTrash
                size={24}
                id={`delete-depositreturn-${index}`}
                title="delete depositreturn"
              />
            }
            onClick={() => original.id && onDelete(original.id)}
          ></Button>
        )}
        {hasClaim(Claims.LEASE_EDIT) && (
          <Button
            title="edit deposit return"
            icon={
              <MdEdit size={24} id={`edit-depositreturn-${index}`} title="edit depositreturn" />
            }
            onClick={() => onEdit(original.id)}
          ></Button>
        )}
      </StyledIcons>
    );
  };
}

export interface IPaymentColumnProps {
  onEdit: (id: number) => void;
  onDelete: (id: number) => void;
}

export const getColumns = ({
  onEdit,
  onDelete,
}: IPaymentColumnProps): ColumnWithProps<ReturnListEntry>[] => {
  return [
    {
      Header: 'Deposit Type',
      accessor: 'depositTypeDescription',
      maxWidth: 50,
    },
    {
      Header: 'Termination or Surrender Date',
      accessor: 'terminationDate',
      align: 'right',
      maxWidth: 50,
      Cell: renderDate,
    },
    {
      Header: 'Deposit amount',
      accessor: 'depositAmount',
      align: 'right',
      maxWidth: 40,
      Cell: renderMoney,
    },
    {
      Header: 'Claims against Deposit',
      accessor: 'claimsAgainst',
      align: 'right',
      maxWidth: 50,
      Cell: renderMoney,
    },
    {
      Header: 'Returned Amount',
      accessor: 'returnAmount',
      align: 'right',
      maxWidth: 50,
      Cell: renderMoney,
    },
    {
      Header: 'Return Date',
      accessor: 'returnDate',
      align: 'right',
      maxWidth: 50,
      Cell: renderDate,
    },

    {
      Header: 'Payee Name',
      accessor: 'personDepositHolder',
      maxWidth: 70,
      Cell: renderPerson,
    },
    {
      Header: 'Actions',
      align: 'right',
      maxWidth: 30,
      Cell: depositActions(onEdit, onDelete),
    },
  ];
};

const StyledIcons = styled(InlineFlexDiv)`
  [id^='edit-depositreturn'] {
    color: ${props => props.theme.css.slideOutBlue};
  }
  [id^='delete-depositreturn'] {
    color: ${props => props.theme.css.discardedColor};
    :hover {
      color: ${({ theme }) => theme.css.dangerColor};
    }
  }
  .btn.btn-primary {
    background-color: transparent;
    padding: 0;
  }
`;
