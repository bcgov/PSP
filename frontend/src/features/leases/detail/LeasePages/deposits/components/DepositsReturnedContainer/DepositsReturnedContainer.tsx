import { FormSection } from 'components/common/form/styles';
import { Table } from 'components/Table';
import { ILeaseSecurityDeposit, ILeaseSecurityDepositReturn } from 'interfaces';

import * as Styled from '../../styles';
import { getColumns, ReturnListEntry } from './columns';

export interface IDepositsReturnedContainerProps {
  securityDeposits: ILeaseSecurityDeposit[];
  depositReturns: ILeaseSecurityDepositReturn[];
  onEdit: (id: number) => void;
  onDelete: (id: number) => void;
}

const DepositsReturnedContainer: React.FC<IDepositsReturnedContainerProps> = ({
  securityDeposits,
  depositReturns,
  onEdit,
  onDelete,
}) => {
  const columns = getColumns({ onEdit, onDelete });
  const dataSource = depositReturns.reduce(
    (accumulator: ReturnListEntry[], returnDeposit: ILeaseSecurityDepositReturn) => {
      var parentDeposit = securityDeposits.find(r => r.id === returnDeposit.parentDepositId);
      if (parentDeposit) {
        accumulator.push(new ReturnListEntry(returnDeposit, parentDeposit));
      }
      return accumulator;
    },
    [],
  );
  return (
    <FormSection>
      <Styled.SectionHeader>Deposits Returned</Styled.SectionHeader>

      <Table<ReturnListEntry>
        name="securityDepositReturnsTable"
        columns={columns}
        data={dataSource}
        manualPagination={false}
        hideToolbar={true}
        noRowsMessage="There is no corresponding data"
      />
    </FormSection>
  );
};

export default DepositsReturnedContainer;
