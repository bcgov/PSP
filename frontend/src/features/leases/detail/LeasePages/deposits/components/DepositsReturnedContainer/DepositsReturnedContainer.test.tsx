import { useKeycloak } from '@react-keycloak/web';
import { ILeaseSecurityDeposit, ILeaseSecurityDepositReturn } from 'interfaces';
import { formatMoney, prettyFormatDate } from 'utils';
import { getAllByRole as getAllByRoleBase, render, RenderOptions } from 'utils/test-utils';

import DepositsReturnedContainer, {
  IDepositsReturnedContainerProps,
} from './DepositsReturnedContainer';

const mockCallback = (id: number): void => {};

const mockDeposit: ILeaseSecurityDeposit = {
  id: 7,
  description: 'Test deposit 1',
  amountPaid: 1234.0,
  depositDate: '2022-02-09',
  depositType: {
    id: 'PET',
    description: 'Pet deposit',
    isDisabled: false,
  },
  rowVersion: 1,
};

const mockDepositReturns: ILeaseSecurityDepositReturn[] = [
  {
    id: 1,
    parentDepositId: 7,
    depositType: {
      id: 'SECURITY',
      description: 'Security deposit',
      isDisabled: false,
    },
    terminationDate: '2022-02-01',
    claimsAgainst: 1234.0,
    returnAmount: 123.0,
    returnDate: '2022-02-16',
    payeeName: '',
    payeeAddress: '',
    rowVersion: 1,
  },
];

jest.mock('@react-keycloak/web');
(useKeycloak as jest.Mock).mockReturnValue({
  keycloak: {
    userInfo: {
      organizations: [1],
      roles: [],
    },
    subject: 'test',
  },
});

// render component under test
const setup = (renderOptions: RenderOptions & IDepositsReturnedContainerProps) => {
  const result = render(
    <DepositsReturnedContainer
      securityDeposits={renderOptions.securityDeposits}
      depositReturns={renderOptions.depositReturns}
      onEdit={mockCallback}
      onDelete={mockCallback}
    />,
    {
      ...renderOptions,
    },
  );

  return {
    ...result,
    findFirstRow: () => {
      const rows = result.getAllByRole('row');
      return rows && rows.length > 1 ? rows[1] : null;
    },
    findCell: (row: HTMLElement, index: number) => {
      const columns = getAllByRoleBase(row, 'cell');
      return columns && columns.length > index ? columns[index] : null;
    },
  };
};

describe('DepositsReturnedContainer component', () => {
  it('renders as expected', () => {
    const { asFragment } = setup({
      securityDeposits: [mockDeposit],
      depositReturns: [...mockDepositReturns],
      onEdit: mockCallback,
      onDelete: mockCallback,
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders one row for each security deposit return', () => {
    const { getAllByRole } = setup({
      securityDeposits: [mockDeposit],
      depositReturns: [...mockDepositReturns],
      onEdit: mockCallback,
      onDelete: mockCallback,
    });
    const rows = getAllByRole('row');
    expect(rows.length).toBe(mockDepositReturns.length + 1);
  });

  it('renders security deposit returns information as expected', () => {
    const depositReturn = mockDepositReturns[0];
    const { findFirstRow, findCell } = setup({
      securityDeposits: [mockDeposit],
      depositReturns: [depositReturn],
      onEdit: mockCallback,
      onDelete: mockCallback,
    });
    const dataRow = findFirstRow() as HTMLElement;

    expect(dataRow).not.toBeNull();
    expect(findCell(dataRow, 0)?.textContent).toBe(depositReturn.depositType.description);
    expect(findCell(dataRow, 1)?.textContent).toBe(prettyFormatDate(depositReturn.terminationDate));
    expect(findCell(dataRow, 2)?.textContent).toBe(formatMoney(mockDeposit.amountPaid));
    expect(findCell(dataRow, 3)?.textContent).toBe(formatMoney(depositReturn.claimsAgainst));
    expect(findCell(dataRow, 4)?.textContent).toBe(formatMoney(depositReturn.returnAmount));
    expect(findCell(dataRow, 5)?.textContent).toBe(prettyFormatDate(depositReturn.returnDate));
  });
});
