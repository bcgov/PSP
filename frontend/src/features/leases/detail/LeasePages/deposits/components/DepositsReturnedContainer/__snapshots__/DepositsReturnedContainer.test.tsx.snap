// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DepositsReturnedContainer component renders as expected 1`] = `
<DocumentFragment>
  <div
    class="Toastify"
  />
  <div
    class="Toastify"
  />
  .c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.c0 {
  border-radius: 1rem;
  width: 100%;
  padding: 2.5rem;
}

.c1 {
  font-size: 1.6rem;
  font-family: 'BCSans-Bold';
}

.c2 {
  font-size: 1.8rem;
  margin-bottom: 2rem;
}

.c4 .btn.btn-primary {
  background-color: transparent;
  padding: 0;
}

<div
    class="c0"
  >
    <label
      class="c1 c2 form-label"
    >
      Deposits Returned
    </label>
    <div
      class="table"
      data-testid="securityDepositReturnsTable"
      role="table"
    >
      <div
        class="thead thead-light"
      >
        <div
          class="tr"
          role="row"
          style="display: flex; flex: 1 0 auto; min-width: 240px;"
        >
          <div
            class="th"
            colspan="1"
            role="columnheader"
            style="box-sizing: border-box; flex: 50 0 auto; min-width: 30px; width: 50px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
          >
            <div
              class="sortable-column"
            >
              Deposit Type
            </div>
          </div>
          <div
            class="th"
            colspan="1"
            role="columnheader"
            style="box-sizing: border-box; flex: 50 0 auto; min-width: 30px; width: 50px; justify-content: right; text-align: right; flex-wrap: wrap; align-items: center; display: flex;"
          >
            <div
              class="sortable-column"
            >
              Termination or Surrender Date
            </div>
          </div>
          <div
            class="th"
            colspan="1"
            role="columnheader"
            style="box-sizing: border-box; flex: 40 0 auto; min-width: 30px; width: 40px; justify-content: right; text-align: right; flex-wrap: wrap; align-items: center; display: flex;"
          >
            <div
              class="sortable-column"
            >
              Deposit amount
            </div>
          </div>
          <div
            class="th"
            colspan="1"
            role="columnheader"
            style="box-sizing: border-box; flex: 50 0 auto; min-width: 30px; width: 50px; justify-content: right; text-align: right; flex-wrap: wrap; align-items: center; display: flex;"
          >
            <div
              class="sortable-column"
            >
              Claims against Deposit
            </div>
          </div>
          <div
            class="th"
            colspan="1"
            role="columnheader"
            style="box-sizing: border-box; flex: 50 0 auto; min-width: 30px; width: 50px; justify-content: right; text-align: right; flex-wrap: wrap; align-items: center; display: flex;"
          >
            <div
              class="sortable-column"
            >
              Returned Amount
            </div>
          </div>
          <div
            class="th"
            colspan="1"
            role="columnheader"
            style="box-sizing: border-box; flex: 50 0 auto; min-width: 30px; width: 50px; justify-content: right; text-align: right; flex-wrap: wrap; align-items: center; display: flex;"
          >
            <div
              class="sortable-column"
            >
              Return Date
            </div>
          </div>
          <div
            class="th"
            colspan="1"
            role="columnheader"
            style="box-sizing: border-box; flex: 70 0 auto; min-width: 30px; width: 70px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
          >
            <div
              class="sortable-column"
            >
              Payee Name
            </div>
          </div>
          <div
            class="th"
            colspan="1"
            role="columnheader"
            style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: right; text-align: right; flex-wrap: wrap; align-items: center; display: flex;"
          >
            <div
              class="sortable-column"
            >
              Actions
            </div>
          </div>
        </div>
      </div>
      <div
        class="tbody"
        role="rowgroup"
      >
        <div
          class="tr-wrapper"
        >
          <div
            class="tr"
            role="row"
            style="display: flex; flex: 1 0 auto; min-width: 240px;"
          >
            <div
              class="td"
              role="cell"
              style="box-sizing: border-box; flex: 50 0 auto; min-width: 30px; width: 50px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
              title=""
            >
              Security deposit
            </div>
            <div
              class="td"
              role="cell"
              style="box-sizing: border-box; flex: 50 0 auto; min-width: 30px; width: 50px; justify-content: right; text-align: right; flex-wrap: wrap; align-items: center; display: flex;"
              title=""
            >
              Feb 1, 2022
            </div>
            <div
              class="td"
              role="cell"
              style="box-sizing: border-box; flex: 40 0 auto; min-width: 30px; width: 40px; justify-content: right; text-align: right; flex-wrap: wrap; align-items: center; display: flex;"
              title=""
            >
              $1,234
            </div>
            <div
              class="td"
              role="cell"
              style="box-sizing: border-box; flex: 50 0 auto; min-width: 30px; width: 50px; justify-content: right; text-align: right; flex-wrap: wrap; align-items: center; display: flex;"
              title=""
            >
              $1,234
            </div>
            <div
              class="td"
              role="cell"
              style="box-sizing: border-box; flex: 50 0 auto; min-width: 30px; width: 50px; justify-content: right; text-align: right; flex-wrap: wrap; align-items: center; display: flex;"
              title=""
            >
              $123
            </div>
            <div
              class="td"
              role="cell"
              style="box-sizing: border-box; flex: 50 0 auto; min-width: 30px; width: 50px; justify-content: right; text-align: right; flex-wrap: wrap; align-items: center; display: flex;"
              title=""
            >
              Feb 16, 2022
            </div>
            <div
              class="td"
              role="cell"
              style="box-sizing: border-box; flex: 70 0 auto; min-width: 30px; width: 70px; justify-content: left; text-align: left; flex-wrap: wrap; align-items: center; display: flex;"
              title=""
            />
            <div
              class="td"
              role="cell"
              style="box-sizing: border-box; flex: 30 0 auto; min-width: 30px; width: 30px; justify-content: right; text-align: right; flex-wrap: wrap; align-items: center; display: flex;"
              title=""
            >
              <div
                class="c3 c4"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
